### DISCUSSION TITLE
Error in Training Tacotron2-DDC

### DISCUSSION THREAD

**Discussion by koayst** [coqui-ai/TTS#1869] [General Q&A]
Hi,

Did I use the wrong command to train Tacotron2-DDC ?

Error => Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu!

Command issued was: CUDA_VISIBLE_DEVICES=0 python train_tacotron_ddc.py

------------------------------

(coqui) xerox@SOIDGXPPDEV01:~/Downloads/TTS/recipes/ljspeech/tacotron2-DDC$ CUDA_VISIBLE_DEVICES=0 python train_tacotron_ddc.py
 > Setting up Audio Processor...
 | > sample_rate:22050
 | > resample:False
 | > num_mels:80
 | > log_func:np.log
 | > min_level_db:-100
 | > frame_shift_ms:None
 | > frame_length_ms:None
 | > ref_level_db:20
 | > fft_size:1024
 | > power:1.5
 | > preemphasis:0.0
 | > griffin_lim_iters:60
 | > signal_norm:False
 | > symmetric_norm:True
 | > mel_fmin:0
 | > mel_fmax:8000
 | > pitch_fmin:0.0
 | > pitch_fmax:640.0
 | > spec_gain:1.0
 | > stft_pad_mode:reflect
 | > max_norm:4.0
 | > clip_norm:True
 | > do_trim_silence:True
 | > trim_db:60.0
 | > do_sound_norm:False
 | > do_amp_to_db_linear:True
 | > do_amp_to_db_mel:True
 | > do_rms_norm:False
 | > db_level:None
 | > stats_path:None
 | > base:2.718281828459045
 | > hop_length:256
 | > win_length:1024
 > Setting up Audio Processor...
 | > sample_rate:22050
 | > resample:False
 | > num_mels:80
 | > log_func:np.log
 | > min_level_db:-100
 | > frame_shift_ms:None
 | > frame_length_ms:None
 | > ref_level_db:20
 | > fft_size:1024
 | > power:1.5
 | > preemphasis:0.0
 | > griffin_lim_iters:60
 | > signal_norm:False
 | > symmetric_norm:True
 | > mel_fmin:0
 | > mel_fmax:8000
 | > pitch_fmin:0.0
 | > pitch_fmax:640.0
 | > spec_gain:1.0
 | > stft_pad_mode:reflect
 | > max_norm:4.0
 | > clip_norm:True
 | > do_trim_silence:True
 | > trim_db:60.0
 | > do_sound_norm:False
 | > do_amp_to_db_linear:True
 | > do_amp_to_db_mel:True
 | > do_rms_norm:False
 | > db_level:None
 | > stats_path:None
 | > base:2.718281828459045
 | > hop_length:256
 | > win_length:1024
 | > Found 13100 files in /home/xerox/Downloads/TTS/recipes/ljspeech/LJSpeech-1.1
 > Using CUDA: True
 > Number of GPUs: 1

 > Model has 47669492 parameters

 > Number of output frames: 6

 > EPOCH: 0/1000
 --> /home/xerox/Downloads/TTS/recipes/ljspeech/tacotron2-DDC/run-August-15-2022_12+57PM-903a77c1
[*] Pre-computing phonemes...
  0%|                                                                                                                 | 0/12969 [00:00<?, ?it/s]fɚ ðɪs pɚpəs d͡ʒɔɹdən weɪtɪd ɔn ðə ɹɪsivɚ tə meɪk ə smɔl peɪmənt, fɚ wɪt͡ʃ hi tɛndɚd ə fɪfti paʊnd noʊt.
 [!] Character '͡' not found in the vocabulary. Discarding it.
wɪt͡ʃ wəz jɛt ʌndɚ fʊl kəntɹoʊl, ænd maɪt bi meɪd tə wɚk kɔɹn mɪlz ɔɹ pɹuv ʌðɚwaɪz pɹədʌktɪv,
 [!] Character '͡' not found in the vocabulary. Discarding it.
æt ðə seɪm steɪʃən pɪɹs d͡ʒɔɪnd eɪɡɚ ɪn ðə ɡɑɹdz væn, ænd ðɛɹ wɚ naʊ θɹi tə kɛɹi ɔn ðə ɹɑbɚi.
 [!] Character '͡' not found in the vocabulary. Discarding it.
ɪt wəz wɛl noʊn ðæt ðeɪ wɚ ɹɛʒɚɛkʃənɪsts, ænd ə stɹɪkt wɔt͡ʃ wəz kɛpt æt ɔl ðə tɚnpaɪk ɡeɪts ɔn ðə sʌðɚn ɹoʊdz lidɪŋ ɪntu lʌndən.
 [!] Character '͡' not found in the vocabulary. Discarding it.
ɪt ɪz sætɪsfæktɚi tə bi eɪbəl tə ɹɪkɔɹd ðæt sʌm kənsɪdɚeɪʃən wəz ʃoʊn ðə kæpɪtəl kɑnvɪkt ækt͡ʃəli əweɪtɪŋ ɛksɪkjuʃən.
 [!] Character '͡' not found in the vocabulary. Discarding it.
ə t͡ʃeɪn fɪkst tə steɪpəlz æt hɪz bæk pæst ɹaʊnd hɪz t͡ʃɛst ʌndɚ hɪz ɑɹmz, ænd wəz pædlɑkt ɔn ðə lɛft saɪd,
 [!] Character '͡' not found in the vocabulary. Discarding it.
ɪnstɛd əv weɪtɪŋ ðɛɹ, ɔzwɔld əpɛɹəntli wɛnt æz fɑɹ əweɪ æz hi kʊd ænd bɔɹdɪd ðə fɚst oʊk klɪf bʌs wɪt͡ʃ keɪm əlɔŋ
 [!] Character '͡' not found in the vocabulary. Discarding it.
ɪd͡ʒwɚ ɹoʊd, kəmplitɪŋ ðɛɹ dɪspəzɪʃənz fɚ əsumɪŋ supɹim paʊɚ æftɚ ðə bloʊ hæd bɪn stɹʌk.
 [!] Character '͡' not found in the vocabulary. Discarding it.
 15%|███████████████                                                                                       | 1912/12969 [01:26<06:55, 26.63it/s]ɪntu ðə “kɹeɪtɚ” dʌɡ aʊt ɪn ðə mɪdəl, pɔɹ ðə spʌnd͡ʒ, wɔɹm wɔtɚ, ðə məlæsɪz, ænd soʊdə dɪzɑlvd ɪn hɑt wɔtɚ.
 [!] Character '“' not found in the vocabulary. Discarding it.
ɪntu ðə “kɹeɪtɚ” dʌɡ aʊt ɪn ðə mɪdəl, pɔɹ ðə spʌnd͡ʒ, wɔɹm wɔtɚ, ðə məlæsɪz, ænd soʊdə dɪzɑlvd ɪn hɑt wɔtɚ.
 [!] Character '”' not found in the vocabulary. Discarding it.
 79%|███████████████████████████████████████████████████████████████████████████████▋                     | 10232/12969 [04:49<01:00, 44.97it/s]ðə fɑsfeɪts wɪt͡ʃ ðə pɹɑsɛs əv “boʊltɪŋ” ɹimuvz tə ə lɑɹd͡ʒ ɪkstɛnt fɹʌm waɪt flaʊɚ, ɡoʊ dɪɹɛktli tə ðə mænjəfækt͡ʃɚ əv boʊn,
 [!] Character '“' not found in the vocabulary. Discarding it.
ðə fɑsfeɪts wɪt͡ʃ ðə pɹɑsɛs əv “boʊltɪŋ” ɹimuvz tə ə lɑɹd͡ʒ ɪkstɛnt fɹʌm waɪt flaʊɚ, ɡoʊ dɪɹɛktli tə ðə mænjəfækt͡ʃɚ əv boʊn,
 [!] Character '”' not found in the vocabulary. Discarding it.
12976it [05:44, 37.64it/s]

> DataLoader initialization
| > Tokenizer:
        | > add_blank: False
        | > use_eos_bos: False
        | > use_phonemes: True
        | > phonemizer:
                | > phoneme language: en-us
                | > phoneme backend: gruut
| > Number of instances : 12969
 | > Preprocessing samples
 | > Max text length: 188
 | > Min text length: 13
 | > Avg text length: 100.90014650319993
 |
 | > Max audio length: 222643.0
 | > Min audio length: 24499.0
 | > Avg audio length: 144984.29755570978
 | > Num. instances discarded samples: 0
 | > Batch group size: 0.

 > TRAINING (2022-08-15 13:02:57)
/home/xerox/Downloads/TTS/TTS/tts/models/tacotron2.py:333: UserWarning: __floordiv__ is deprecated, and its behavior will change in a future version of pytorch. It currently rounds toward 0 (like the 'trunc' function NOT 'floor'). This results in incorrect rounding for negative values. To keep the current behavior, use torch.div(a, b, rounding_mode='trunc'), or for actual floor division, use torch.div(a, b, rounding_mode='floor').
  ) // self.decoder.r
 ! Run is removed from /home/xerox/Downloads/TTS/recipes/ljspeech/tacotron2-DDC/run-August-15-2022_12+57PM-903a77c1
Traceback (most recent call last):
  File "/home/xerox/anaconda3/envs/coqui/lib/python3.7/site-packages/trainer/trainer.py", line 1533, in fit
    self._fit()
  File "/home/xerox/anaconda3/envs/coqui/lib/python3.7/site-packages/trainer/trainer.py", line 1517, in _fit
    self.train_epoch()
  File "/home/xerox/anaconda3/envs/coqui/lib/python3.7/site-packages/trainer/trainer.py", line 1282, in train_epoch
    _, _ = self.train_step(batch, batch_num_steps, cur_step, loader_start_time)
  File "/home/xerox/anaconda3/envs/coqui/lib/python3.7/site-packages/trainer/trainer.py", line 1123, in train_step
    num_optimizers=1,
  File "/home/xerox/anaconda3/envs/coqui/lib/python3.7/site-packages/trainer/trainer.py", line 998, in _optimize
    outputs, loss_dict = self._model_train_step(batch, model, criterion)
  File "/home/xerox/anaconda3/envs/coqui/lib/python3.7/site-packages/trainer/trainer.py", line 954, in _model_train_step
    return model.train_step(*input_args)
  File "/home/xerox/Downloads/TTS/TTS/tts/models/tacotron2.py", line 353, in train_step
    text_lengths,
  File "/home/xerox/anaconda3/envs/coqui/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/xerox/Downloads/TTS/TTS/tts/layers/losses.py", line 441, in forward
    if self.config.stopnet
  File "/home/xerox/anaconda3/envs/coqui/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/xerox/Downloads/TTS/TTS/tts/layers/losses.py", line 194, in forward
    x.masked_select(mask), target.masked_select(mask), pos_weight=self.pos_weight, reduction="sum"
  File "/home/xerox/anaconda3/envs/coqui/lib/python3.7/site-packages/torch/nn/functional.py", line 3150, in binary_cross_entropy_with_logits
    return torch.binary_cross_entropy_with_logits(input, target, weight, pos_weight, reduction_enum)
RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu!

**Comment by erogol** [comment]
It looks like a bug. Checking it.

**Reply by koayst** [reply]
Hi @erogol,

I tried to train " tacotron2-Capacitron" under recipes and also encountered the same error.  

Error => RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu!

I wonder whether it is a related "issue".

Thanks for checking.

**Reply by erogol** [reply]
Fixed here https://github.com/coqui-ai/TTS/pull/1872

**Reply by koayst** [reply]
Hi @erogol 

How do I get the changes ?  What is the git command ?  

Or do I manually change it over my end as I noticed there is one file changed in TTS/tts/layers/losses.py ?

**Comment by koayst** [comment]
I pulled all the latest changes and tried to train.  It seems to be training all right. now.  Now I just have to wait for the training to complete and then test it out.
